### PR TITLE
Fix various problems with IN-GJ

### LINKED ISSUE #1298: IN-GJ (Gujarat) parser needs improvement
I was about to add capacity data to **IN-GJ** and looked up the data source from the parser for conventional generation.
https://www.sldcguj.com/RealTimeData/RealTimeDemand.php
It also has capacity data for the units of the power plants.

1. While cross checking with the map, I found it highly confusing that "Gujarat Total Conventional Generation" on that site (about 8500 MW) was above the total value for all combined types in the origin graph.

image

2. GMR is mapped to wind, but should be something thermal PP as well (I suggest to "unknown" - the wind numbers are in a seperate figure).

3. CGPL is mapped to coal, with a current generation of 2530 MW, which is totally missing on the map.
I assume the two spaces "  " behind "CGPL  " may cause this trouble.
This is a huge amount that is missing in the "coal" category and would fill up a large part of the gap.

4. The "unknown" category could be filled with what is mapped as "thermal" in the parser. That figure is definitely missing on the map. (~2800 MW)

5. Here is a screenshot of the recent mapping. I colored the fields according to their recent mapping. I also edited the fields with "123 + 456" to be the sum of both (generating units)

image

### PR DESCRIPTION
**Pull request by systemcatch**
- Use @alixunderplatz's new station mapping.
- More reliable regex substitution.
- Add specific ignore for DSM Rate which was causing failures.
- Better fetch_production docstring.
- Add nuclear key.
- Couple of variable names changed for clarity.
- Make sure new unknown plants are logged even if literal_eval fails.

fixes #1298

### CODE REVIEW

**Comment by alixunderplatz** [comment]
@systemcatch Lovely! I guess you could fix that "whitespace issue" that caused some generation to get lost, also with the new mapping?

**Comment by systemcatch** [comment]
The new regex gets rid of all spaces in the plant names unlike the old one which only got some. I modified your station mapping to compensate for this.

**Comment by systemcatch** [comment]
@maxbellec I suspected you wouldn't be able to resist doing some extra cleaning :laughing:, does the parser still run ok?

**Comment by maxbellec** [comment]
yes, I've tested the changes 
```python
In [1]: from parsers.IN_GJ import fetch_production, fetch_consumption

In [2]: fetch_consumption()
Out[2]: 
{'consumption': 15010.0,
 'datetime': 0,
 'source': 'sldcguj.com',
 'zoneKey': 'IN-GJ'}

In [3]: fetch_production()
Out[3]: 
{'datetime': 0,
 'production': {'biomass': None,
  'coal': 4347.0,
  'gas': 1500.0,
  'geothermal': None,
  'hydro': 78.0,
  'nuclear': 0,
  'oil': None,
  'solar': 576.0,
  'unknown': 0,
  'wind': 1341.0},
 'source': 'sldcguj.com',
 'storage': {'hydro': None},
 'zoneKey': 'IN-GJ'}
```

**Comment by alixunderplatz** [comment]
@systemcatch Does the parser take into account negative values (self-consumption of inactive units)? If it does, that's a minor inaccuracy because it would lower the output of a certain generation type and lower emissions in return.

@maxbellec I compared your numbers from the test to adding the numbers from the website manually according to my mapping (40 minutes difference).
There still were crucial differences for gas and coal. I assume, some of the values are still omitted. 

@systemcatch Is there an issue with that annoying "generation" column? When I copy the chart into a spreadsheet, I get plenty of whitespaces behind some numbers as well and have to fix them and the plus signs etc. manually.

coal: 5814 (1 600 MW more)
gas: 2041 (500 MW more)
hydro: 80 (fine)
solar: 464 (fine)
wind: 1526 (fine)

**Comment by maxbellec** [comment]
I have not (or at least not volontarily, but I double-checked and I'm quite confident) done any functional change to systemcatch's code. I'll try to compare the three.

**Comment by maxbellec** [comment]
- latest changes
```python
In [1]: from parsers.IN_GJ import fetch_production, fetch_consumption

In [2]: fetch_consumption()
Out[2]: 
{'consumption': 13878.0,
 'datetime': 0,
 'source': 'sldcguj.com',
 'zoneKey': 'IN-GJ'}

In [3]: fetch_production()
hydro returned -2.00, setting to None
Out[3]: 
{'datetime': 0,
 'production': {'biomass': None,
  'coal': 3870.0,
  'gas': 1204.0,
  'geothermal': None,
  'hydro': None,
  'nuclear': 0,
  'oil': None,
  'solar': 5.0,
  'unknown': 0,
  'wind': 1750.0},
 'source': 'sldcguj.com',
 'storage': {'hydro': None},
 'zoneKey': 'IN-GJ'}
```

- Latest systemcatch code:
```python
In [1]: from parsers.IN_GJ import fetch_production, fetch_consumption

In [2]: fetch_consumption()
Out[2]: 
{'consumption': 13878.0,
 'datetime': datetime.datetime(2018, 4, 25, 19, 20, 50, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Kolkata')),
 'source': 'sldcguj.com',
 'zoneKey': 'IN-GJ'}

In [3]: fetch_production()
hydro returned -2.00, setting to None
Out[3]: 
{'datetime': datetime.datetime(2018, 4, 25, 19, 21, 50, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Kolkata')),
 'production': {'biomass': None,
  'coal': 3875.0,
  'gas': 1204.0,
  'geothermal': None,
  'hydro': None,
  'nuclear': 0.0,
  'oil': None,
  'solar': 5.0,
  'unknown': 0.0,
  'wind': 1752.0},
 'source': 'sldcguj.com',
 'storage': {'hydro': None},
 'zoneKey': 'IN-GJ'}
```

- previous code: just failed. Why?

**Comment by systemcatch** [comment]
@alixunderplatz 

In answer to the first question I ran the parser modified to display each value before they are passed to eval() and it looks like small negative values are being counted.

```
(EM-env) chris@ThinkPad:~/electricitymap$ python -m parsers.IN_GJ
Ukai(1-5)+Ukai6
612+-1
Wanakbori
1449
Gandhinagar
528
Sikka(1-2)+Sikka(3-4)
0+369
KLTPS(1-3)+KLTPS4
61+0
Utran(Gas)(II)
-3
Dhuvaran(Gas)(I)+(II)
66+0
Ukai(Hydro)
0
Kadana(Hydro)
0
GIPCL(I)
92
SLPP(I+II)
240+231
Akrimota
203
GSEG(I+II)
-3+-1
GPPC
243
CLPI
-30
TPAECo
357
EPGL(I+II)
0
Sugen+Unosgn
900
Adani(I+II+III)
163+0+0
BECL(I+II)
0+0
SSP(RBPH)
-2
hydro returned -2.00, setting to None
{'datetime': datetime.datetime(2018, 4, 25, 20, 27, 50, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Kolkata')),
 'production': {'biomass': None,
                'coal': 4212.0,
                'gas': 1264.0,
                'geothermal': None,
                'hydro': None,
                'nuclear': 0,
                'oil': None,
                'solar': 5.0,
                'unknown': 0,
                'wind': 1611.0},
 'source': 'sldcguj.com',
 'storage': {'hydro': None},
 'zoneKey': 'IN-GJ'}
```
The generation column has all whitespace removed from it before it's passed to eval().

**Comment by maxbellec** [comment]
as @alixunderplatz found out, there's an issue in the mapping. If I print `(v1, v2)` each time and the corresponding mapping, we see some information is missing:

```
v1=StateLoadDespatchCentre(SLDC)-Real-timeSystemData.style1{color:#0000FF;}.style2{font-family:Arial;font-size:x-small;}.style3{font-family:Arial;font-size:x-small;color:#FFFFFF;}.style6{font-family:Arial;font-size:small;}.style7{font-family:Arial;font-size:small;color:#0000FF;}.style8{font-family:Arial;font-size:small;color:#FFFFFF;}.style10{font-size:medium;}.style11{background-color:#BCD9DC;}.style15{border-style:solid;border-width:0;background-color:#336699;}.style16{font-family:Arial;font-size:x-small;color:#0000FF;}.style17{text-align:center;}.style18{text-align:center;background-color:#BCD9DC;}.style19{text-align:center;border-style:solid;border-width:0;background-color:#336699;}.style20{font-family:Arial;font-size:x-small;color:#FF0000;}.style21{font-size:medium;color:#FF0000;}.style22{font-family:Arial;}.style23{font-size:small;color:#FF0000;}.style24{font-size:small;color:#FF0000;font-family:Arial;}.style25{font-size:small;font-family:Arial;color:#FF00FF;border-style:none;border-width:medium;background-color:#BCD9DC;}.auto-style1{border-style:solid;border-width:0;}.auto-style3{font-size:small;font-family:Arial;color:#FF00FF;border-style:solid;border-width:0;background-color:#BCD9DC;}.auto-style5{border-style:solid;border-width:0;font-size:small;color:#FF0000;font-family:Arial;}.auto-style6{text-align:center;border-style:solid;border-width:0;}.auto-style7{border-width:1;}GujaratPowerSystem-Real-timeData25-04-201820:44:5049.996HzDSMRate(Rs./Unit):2.138StationNameInstalledCapacityScheduleGenerationNameDrawalScheduleDSMUkai(1-5)+Ukai6850+500556,0600+-1GujaratAtperiphery65676309258Wanakbori14701148,1811451DGVCL28712576293Gandhinagar630384,194512MGVCL1674159571Sikka(1-2)+Sikka(3-4)240+5403800+368UGVCL31712978161KLTPS(1-3)+KLTPS4215+7556,064+0PGVCL37004131-439Utran(Gas)(II)3750-3AECo+SECo15451620-75Dhuvaran(Gas)(I)+(II)219+3760,066+0Ukai(Hydro)305-1Kadana(Hydro)2420GSECLTotalGeneration3054GIPCL(I)145(52)2892PublicownedIPPGeneration766SLPP(I+II)250+250215,198242+232PrivateIPPGeneration134Akrimota250174203TorrentPower1185GSEG(I+II)156+3510-3+-1GujaratTotalConventionalGeneration7232GPPC7030243(Wind+Solar)Generation1604+5CLPI6550+0-31TPAECo500357GujaratCateredMW13799EPGL(I+II)1200(1000)00Sugen+Unosgn1148+383835828Adani(I+II+III)1320(1000)+1320(1000)+1980(0)162166+0+0BECL(I+II)250+25000+0CGPL4150(1971)2920, v2=DSMRate(Rs./Unit):2.138
v1=StationName, v2=Generation
v1=Ukai(1-5)+Ukai6, v2=600+-1
ok, correspond to ['coal']
v1=Wanakbori, v2=1451
ok, correspond to ['coal']
v1=Gandhinagar, v2=512
ok, correspond to ['coal']
v1=Sikka(1-2)+Sikka(3-4), v2=0+368
ok, correspond to ['coal']
v1=KLTPS(1-3)+KLTPS4, v2=64+0
ok, correspond to ['coal']
v1=Utran(Gas)(II), v2=-3
ok, correspond to ['gas']
v1=Dhuvaran(Gas)(I)+(II), v2=66+0
ok, correspond to ['gas']
v1=Ukai(Hydro), v2=-1
ok, correspond to ['hydro']
v1=Kadana(Hydro), v2=0
ok, correspond to ['hydro']
v1=GIPCL(I), v2=92
ok, correspond to ['gas']
v1=SLPP(I+II), v2=242+232
ok, correspond to ['coal']
v1=Akrimota, v2=203
ok, correspond to ['coal']
v1=GSEG(I+II), v2=-3+-1
ok, correspond to ['gas']
v1=GPPC, v2=243
ok, correspond to ['gas']
v1=CLPI, v2=-31
ok, correspond to ['gas']
v1=TPAECo, v2=357
ok, correspond to ['coal']
v1=EPGL(I+II), v2=0
ok, correspond to ['coal']
v1=Sugen+Unosgn, v2=828
ok, correspond to ['gas']
v1=Adani(I+II+III), v2=166+0+0
ok, correspond to ['coal']
v1=BECL(I+II), v2=0+0
ok, correspond to ['coal']
v1=SSP(RBPH), v2=-2
ok, correspond to ['hydro']
```

**Comment by maxbellec** [comment]
for example CGPL, KAPP, KAWAS are ignored
screenshot at 2018-04-25 17-14-32

**Comment by systemcatch** [comment]
CGPL, KAPP, KAWAS are all in the new mapping but don't get assigned to the correct type. Even worse they aren't assigned to unknown and logged but end up being dropped silently.

**Comment by systemcatch** [comment]
Ok so I get what is going on now. The parser gets the len of https://github.com/tmrowco/electricitymap/pull/1349/files?diff=split#diff-9c14902f5522fe35ae5350091827a1d2R78 to figure out if a row is production (>3) or maybe consumption (=3) but crucially it doesn't check for len < 3. CGPL/KAPP/KAWAS/JHANOR/GMR all have len=1 for some reason so are totally ignored.

I will add in a specific check for these plants.

File: parsers/IN_GJ.py [hunk 1]
```diff
@@ -45,6 +46,19 @@ def to_float(s):
...
 
+def split_and_sum(expression):
+    """
+    Avoid using literal_eval for simple addition expressions.
+    Returns sum of all positive numbers.
+    """
+
+    split_vals = expression.split('+')
+    float_vals = [float(v) for v in split_vals]
+    total = sum([v for v in float_vals if v > 0.0])
```
Comment (maxbellec) [inline parsers/IN_GJ.py]: why keep only positive numbers?

Comment (maxbellec) [reply parsers/IN_GJ.py]: can you not replace all `to_float` by `split_and_sum`?

Comment (alixunderplatz) [reply parsers/IN_GJ.py]: @maxbellec the inactive units have a bit of self-consumption, which causes the negative values in the chart. One of them had like -30 MW yesterday, which would reduce the output of coal by the same amount, and therefore total emissions, too, which would be slightly inaccurate.

Comment (systemcatch) [reply parsers/IN_GJ.py]: The negative numbers are not correct and including them in the sum reduces the generation for that fuel type. I can replace to_float entirely if you're happy with that.

Comment (maxbellec) [reply parsers/IN_GJ.py]: alright! Yeah, replace all `to_float` by `split_and_sum` seems better, I was not a big fan of `literal_eval` either

**Comment by systemcatch** [comment]
@maxbellec @alixunderplatz can we merge this if there's nothing else that needs doing.

**Comment by alixunderplatz** [comment]
@systemcatch I updated the branch and now the checks have failed for some reason...

**Comment by maxbellec** [comment]
restarting tests fixed it...

**Comment by maxbellec** [comment]
great work on this @systemcatch and @alixunderplatz !

**Pull request merged**